pkg/controller/node/ipam/cloud_cidr_allocator.go

```diff
@@ -42,84 +45,149 @@ import (
-	glog.V(2).Infof("Updating PodCIDR for node %v", node.Name)
+func (ca *cloudCIDRAllocator) removeNodeFromProcessing(nodeName string) {
+	ca.Lock()
+	defer ca.Unlock()
+	ca.nodesInProcessing.Delete(nodeName)
+}
 
+// WARNING: If you're adding any return calls or defer any more work from this
+// function you have to handle correctly nodesInProcessing.
+func (ca *cloudCIDRAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	if node == nil {
+		return nil
+	}
+	if !ca.insertNodeToProcessing(node.Name) {
+		glog.V(2).Infof("Node %v is already in a process of CIDR assignment.", node.Name)
+		return nil
+	}
 	cidrs, err := ca.cloud.AliasRanges(types.NodeName(node.Name))
-
 	if err != nil {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-
 	if len(cidrs) == 0 {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
-		glog.V(2).Infof("Node %v has no CIDRs", node.Name)
-		return fmt.Errorf("failed to allocate cidr (none exist)")
+		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
 
-	node, err = ca.client.Core().Nodes().Get(node.Name, metav1.GetOptions{})
-	if err != nil {
-		glog.Errorf("Could not get Node object from Kubernetes: %v", err)
-		return err
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, cidrs[0])
+	ca.nodeCIDRUpdateChannel <- nodeAndCIDR{
+		nodeName: node.Name,
+		cidr:     cidrs[0],
 	}
+	return nil
+}
 
-	podCIDR := cidrs[0]
-
-	if node.Spec.PodCIDR != "" {
-		if node.Spec.PodCIDR == podCIDR {
-			glog.V(3).Infof("Node %v has PodCIDR %v", node.Name, podCIDR)
-			return nil
+// Assigns CIDR to Node and sends an update to the API server.
+func (ca *cloudCIDRAllocator) updateCIDRAllocation(data nodeAndCIDR) error {
+	var err error
+	var node *v1.Node
+	defer ca.removeNodeFromProcessing(data.nodeName)
+	podCIDR := data.cidr
+	for rep := 0; rep < cidrUpdateRetries; rep++ {
+		// TODO: change it to using PATCH instead of full Node updates.
+		node, err = ca.client.Core().Nodes().Get(data.nodeName, metav1.GetOptions{})
+		if err != nil {
+			glog.Errorf("Failed while getting node %v to retry updating Node.Spec.PodCIDR: %v", data.nodeName, err)
+			continue
+		}
+		if node.Spec.PodCIDR != "" {
+			if node.Spec.PodCIDR == podCIDR {
+				glog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+				return nil
+			}
+			glog.Errorf("PodCIDR cannot be reassigned, node %v spec has %v, but cloud provider has assigned %v",
+				node.Name, node.Spec.PodCIDR, podCIDR)
+			// We fall through and set the CIDR despite this error. This
+			// implements the same logic as implemented in the
+			// rangeAllocator.
+			//
+			// See https://github.com/kubernetes/kubernetes/pull/42147#discussion_r103357248
```

Comment:
Feels like this should `break`, otherwise it will just spin the loop.

Reply:
It wasn't before, but I think it should return, right?

Reply:
> Feels like this should break, otherwise it will just spin the loop.

I don't think it'll spin the loop. It will just go outside the if condition where we're updating the node's CIDR to the given one and then break only if the update succeeded (otherwise loop back). Seems to be the intended behavior. Am I missing something?

> It wasn't before, but I think it should return, right?

I think it shouldn't return here, because you need to go further below (outside the loop) where the 'RouteCreated' node condition is set.

Reply:
in this case `if node.Spec.PodCIDR != podCIDR`

We can't update the node Spec actually after it is set, so there is no point in retrying? That's why it should return...

Reply:
I thought the intended behavior there is to change the node spec (even if it had a different value before) if the cloud provider changed the ip-range? The error msg was misleading earlier, saying `PodCIDR cannot be reassigned...`, I changed it to `PodCIDR being reassigned! ...`. Does it sgty?

Reply:
If you try to edit a PodCIDR after it has been assign, the API server will now reject the update request with a validation error. At least that is what happens when I edit the object via kubectl.

Reply:
Hmm.. that's interesting. So does that mean each node requires processing by the cloud-cidr-allocator only once during the cluster's entire lifetime?
Btw - this behavior of trying to go ahead and still set the new podCIDR is what was there before. I just left it the same way in my PR.